File: 2251-number-of-flowers-in-full-bloom/2251-number-of-flowers-in-full-bloom.py
class Solution:
    def fullBloomFlowers(self, flowers: List[List[int]], people: List[int]) -> List[int]:
        
        starts, ends = [], []
        n = len(flowers)
        for start, end in flowers:
            starts.append(start)
            ends.append(end)
        
        starts.sort()
        ends.sort()
        ans = []
        for p in people:
            start = bisect_right(starts, p)
            end = bisect_left(ends, p)
            ans.append(start- end)
        return ans

File: 2251-number-of-flowers-in-full-bloom/2251-number-of-flowers-in-full-bloom.cpp
class Solution {
public:
    vector<int> fullBloomFlowers(vector<vector<int>>& flowers, vector<int>& people) {
        vector<int> init_state(people.begin(), people.end());
        sort(people.begin(), people.end());
        sort(flowers.begin(), flowers.end());
        priority_queue<int, vector<int>, greater<int> > pq;
        
        unordered_map<int, int> blooms;
        int cur_flower = 0;
        int n = people.size();
        for (int i = 0; i <  people.size(); ++i) {
            int arrival_time = people[i];
            
            while (cur_flower < flowers.size() and flowers[cur_flower][0] <= arrival_time) 
                pq.push(flowers[cur_flower++][1]);
            
            while (!pq.empty() and pq.top() < arrival_time) 
                pq.pop();
                
            blooms[arrival_time] = (pq.size());
        }
        vector<int> ans;
        for (int i = 0; i < init_state.size(); ++i) {
            ans.push_back(blooms[init_state[i]]);
        }
        return ans;
    }
};